### PR TITLE
ARROW-7288: alternate R-only workaround for std::regex issue

### PR DESCRIPTION
**Pull request by nealrichardson**

### CODE REVIEW

**Comment by jeroen** [comment]
Few comments: 

 - You can use `withr::with_locale()` to simplify temporary setting the locale
 - Most servers won't have the Japanese language pack installed, so `Sys.setlocale("LC_COLLATE", "Japanese")` may fail
 - The bug is that the action takes very long, but it doesn't crash. So I'm not sure you'll detect the problem properly, or just that your tests will be taking super long to complete.

File: r/R/parquet.R [hunk 1]
```diff
@@ -567,3 +568,14 @@ ParquetArrowReaderProperties <- R6Class("ParquetArrowReaderProperties",
...
   parquet___arrow___ArrowReaderProperties__Make(isTRUE(use_threads))
 }
+
+with_safe_locale <- function(expr) {
+  # ARROW-7288: due to a bug in libstdc++ on mingw, std::regex can blow up
+  # on other locales, and the parquet reader calls it
+  if (tolower(Sys.info()[["sysname"]]) == "windows") {
+    old_locale <- Sys.getlocale("LC_COLLATE")
+    Sys.setlocale("LC_COLLATE", "C")
+    on.exit(Sys.setlocale(old_locale))
```
Comment (yutannihilation) [inline r/R/parquet.R:578]:
```suggestion
    on.exit(Sys.setlocale("LC_COLLATE", old_locale))
```

`category` is needed here, as the error message on CI suggests.

```
Error: Error: invalid 'category' argument
```

**Review by yutannihilation — COMMENTED** [review]
I agree withr is a handy choice. It imports only base packages and well-maintained.

Just in case we decide not to use it, one minor comment.

**Comment by nealrichardson** [comment]
Thanks both. Agree that withr is handy but didn't think it worth adding a dependency (even a light one) for something so trivial. In any case, it looks like we'll solve this more properly (either #9320 or #9332), and as Jeroen points out, it's probably not a safe thing to have a test for anyway, so I'll close this.

**Pull request closed without merging**